<a href="https://colab.research.google.com/github/devadathen/datasciencelab/blob/main/Devadathan_U_KNN3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#### Goal of the Project

This project is designed for you to practice and solve the activities that are based on the concept:  kNN.




---

### Problem Statement

As an owner of a startup, you wish to forecast the sales of your product to plan how much money should be spent on advertisements. This is because the sale of a product is usually proportional to the money spent on advertisements. To analyse this, you are given a dataset having the following attributes:         

|Attribute|Description|
|-|-|
|TV| TV advertising budget in thousands of dollars.|
|Radio| Radio advertising budget in thousands of dollars.|
|Newspaper| Newspaper advertising budget in thousands of dollars.|
|Sales| Product Sales in thousands of dollars.|

  **Source:** https://www.kaggle.com/ishaanv/ISLR-Auto

Predict the impact of TV advertising on your product sales by using kNN regression and evaluate the accuracy of the model.





---

### List of Activities

**Activity 1:** Import Modules and Read Data
  
**Activity 2:**  Perform Train-Test Split

**Activity 4:** Build kNN Regressor Model






---


#### Activity 1: Import Modules and Read Data

Create a Pandas DataFrame for **Advertising-Sales** dataset using the below link. This dataset contains information about the money spent on the TV, radio, and newspaper advertisement (in thousand dollars) and their generated sales (in thousand units). The dataset consists of examples that are divided by 1000.



  **Dataset :** advertising.csv

Also, print the first five rows of the dataset. Check for null values and treat them accordingly.




In [ ]:
# Import modules
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score
import pandas as pd
# Load the dataset
df = pd.read_csv('/content/advertising.csv')
# Print first five rows using head() function
df.head(50)

TV  Radio  Newspaper  Sales
0   230.1   37.8       69.2   22.1
1    44.5   39.3       45.1   10.4
2    17.2   45.9       69.3   12.0
3   151.5   41.3       58.5   16.5
4   180.8   10.8       58.4   17.9
5     8.7   48.9       75.0    7.2
6    57.5   32.8       23.5   11.8
7   120.2   19.6       11.6   13.2
8     8.6    2.1        1.0    4.8
9   199.8    2.6       21.2   15.6
10   66.1    5.8       24.2   12.6
11  214.7   24.0        4.0   17.4
12   23.8   35.1       65.9    9.2
13   97.5    7.6        7.2   13.7
14  204.1   32.9       46.0   19.0
15  195.4   47.7       52.9   22.4
16   67.8   36.6      114.0   12.5
17  281.4   39.6       55.8   24.4
18   69.2   20.5       18.3   11.3
19  147.3   23.9       19.1   14.6
20  218.4   27.7       53.4   18.0
21  237.4    5.1       23.5   17.5
22   13.2   15.9       49.6    5.6
23  228.3   16.9       26.2   20.5
24   62.3   12.6       18.3    9.7
25  262.9    3.5       19.5   17.0
26  142.9   29.3       12.6   15.0
27  240.1   16.7       22.9   20.9
28  248.8   27.1       22.9   18.9
29   70.6   16.0       40.8   10.5
30  292.9   28.3       43.2   21.4
31  112.9   17.4       38.6   11.9
32   97.2    1.5       30.0   13.2
33  265.6   20.0        0.3   17.4
34   95.7    1.4        7.4   11.9
35  290.7    4.1        8.5   17.8
36  266.9   43.8        5.0   25.4
37   74.7   49.4       45.7   14.7
38   43.1   26.7       35.1   10.1
39  228.0   37.7       32.0   21.5
40  202.5   22.3       31.6   16.6
41  177.0   33.4       38.7   17.1
42  293.6   27.7        1.8   20.7
43  206.9    8.4       26.4   17.9
44   25.1   25.7       43.3    8.5
45  175.1   22.5       31.5   16.1
46   89.7    9.9       35.7   10.6
47  239.9   41.5       18.5   23.2
48  227.2   15.8       49.9   19.8
49   66.9   11.7       36.8    9.7

In [ ]:
# Check if there are any null values. If any column has null values, treat them accordingly.
df.isnull().sum()

TV           0
Radio        0
Newspaper    0
Sales        0
dtype: int64

**Q:** Are there any missing or null values in the dataset?

**A:**no


---

#### Activity 2: Perform Train-Test Split

In this dataset, `Sales` is the target variable and all other columns other than `Sales` are feature variables.

Create two separate DataFrames, one containing the feature variables and the other containing the target variable.





In [ ]:
df.columns

Index(['TV', 'Radio', 'Newspaper', 'Sales'], dtype='object')

In [ ]:
x=df[['TV', 'Radio', 'Newspaper', 'Sales']]
y=df['Sales']

In [ ]:
# Split the dataset into dependent and independent features
feature_df=df.drop(['Sales'],axis=1)
target_df=df['Sales']

Normalise all the feature variables using the `StandardScaler` technique so that all the features have mean `0` and the same variance before applying kNN.

In [ ]:
# Normalise the feature variables using 'StandardScaler'.

# Import 'StandardScaler' from 'sklearn.preprocessing' module.
from sklearn.preprocessing import StandardScaler
# Create an object of 'StandardScaler' and call 'fit_transform()' function by passing feature variables.
ob = StandardScaler()
scaled_df = ob.fit_transform(feature_df)
# print(scaled_df)
# Convert the scaled features array obtained from 'fit_transform()' function into a DataFrame.
converted_df = pd.DataFrame(scaled_df)
converted_df.columns = feature_df.columns
converted_df.head(50)

TV     Radio  Newspaper
0   0.969852  0.981522   1.778945
1  -1.197376  1.082808   0.669579
2  -1.516155  1.528463   1.783549
3   0.052050  1.217855   1.286405
4   0.394182 -0.841614   1.281802
5  -1.615408  1.731034   2.045930
6  -1.045577  0.643905  -0.324708
7  -0.313437 -0.247406  -0.872487
8  -1.616576 -1.429069  -1.360424
9   0.616043 -1.395307  -0.430582
10 -0.945156 -1.179231  -0.292486
11  0.790028  0.049697  -1.222329
12 -1.439088  0.799209   1.627040
13 -0.578502 -1.057689  -1.075027
14  0.666253  0.650657   0.711007
15  0.564665  1.650006   1.028627
16 -0.925305  0.900494   3.841171
17  1.568876  1.103065   1.162119
18 -0.908957 -0.186635  -0.564074
19  0.003007  0.042945  -0.527248
20  0.833233  0.299535   1.051643
21  1.055093 -1.226498  -0.324708
22 -1.562862 -0.497243   0.876722
23  0.948834 -0.429720  -0.200423
24 -0.989528 -0.720071  -0.564074
25  1.352854 -1.334536  -0.508836
26 -0.048371  0.407572  -0.826455
27  1.086621 -0.443225  -0.352328
28  1.188210  0.259020  -0.352328
29 -0.892610 -0.490491   0.471642
30  1.703160  0.340049   0.582118
31 -0.398678 -0.395958   0.370372
32 -0.582005 -1.469583  -0.025502
33  1.384381 -0.220397  -1.392646
34 -0.599520 -1.476335  -1.065821
35  1.677471 -1.294022  -1.015186
36  1.399561  1.386664  -1.176297
37 -0.844735  1.764796   0.697198
38 -1.213724  0.232011   0.209261
39  0.945331  0.974770   0.066562
40  0.647570 -0.065093   0.048149
41  0.349810  0.684419   0.374975
42  1.711334  0.299535  -1.323599
43  0.698949 -1.003670  -0.191216
44 -1.423908  0.164487   0.586721
45  0.327624 -0.051588   0.043546
46 -0.669581 -0.902385   0.236880
47  1.084286  1.231360  -0.554867
48  0.935989 -0.503996   0.890531
49 -0.935814 -0.780842   0.287515

Split the dataset into a train set and test set such that the train set contains 70% of the instances and the remaining instances will become the test set.

In [ ]:
# Split the DataFrame into the train and test sets.
# Perform train-test split using 'train_test_split' function.
x_train,x_test,y_train,y_test=train_test_split(converted_df,target_df,train_size=0.7,random_state=1)

# Print the shape of train and test sets.
print("x Train",x_train.shape)
print("x Test",x_test.shape)
print("y Train",y_train.shape)
print("y Test",y_test.shape)

x Train (140, 3)
x Test (60, 3)
y Train (140,)
y Test (60,)


After this activity, you must obtain train and test sets so that they can be used for training and testing the kNN regressor model.

---

#### Activity 3: Build kNN Regressor Model

Deploy the kNN regressor model for the optimal value of $k$ using the steps given below:   

1. Import the `KNeighborsRegressor` class from the `sklearn.neighbors` module (if not imported yet).

2. Create an object of `KNeighborsRegressor` and pass the optimal $k$ value 2 as input to its constructor.

3. Call the `fit()` function using the regressor object and pass the train set as inputs to this function.

4. Perform prediction for train and test sets using the `predict()` function.

5. Also, determine the accuracy score of the train and test sets using the `score()` function.

In [ ]:
from sklearn.neighbors import KNeighborsRegressor
# Train kNN regressor model
kn = KNeighborsRegressor(n_neighbors=2)
kn.fit(x_train,y_train)
# Perform prediction using 'predict()' function.
y_predicted = kn.predict(x_test)
sample_data = [[-1.516155 ,1.528463,1.783549]]
p=kn.predict(sample_data)
print("Predicted Values=",p)
print("Test score",kn.score(x_test,y_test))
# Call the 'score()' function to check the accuracy score of the train set and test set.
print("Train score",kn.score(x_train,y_train))

Predicted Values= [9.6]
Test score 0.9243088142911436
Train score 0.9635754264321423


/usr/local/lib/python3.10/dist-packages/sklearn/base.py:439: UserWarning: X does not have valid feature names, but KNeighborsRegressor was fitted with feature names
  warnings.warn(


**Q:** Write down the train and test set accuracy scores for the kNN regressor?

Test score 0.9243088142911436
Train score 0.9635754264321423


After this activity, you must obtain a kNN regressor model using the `sklearn` module for predicting total sales based on advertising budgets.

---